feat(provider): add Robots.txt provider for extracting disallowed and…

Adesoji1 opened the pull request:
I Added Robots.txt Provider to Discover URLs from Disallow and Sitemap Directives

hahwul commented on src/providers/robots.rs:
```diff
@@ -0,0 +1,117 @@
+            let mut urls = Vec::new();
+
+            let resp = client.get(&url).send().await?;
+            if !resp.status().is_success() {
+                return Ok(urls);
+            }
```
No modification plan

hahwul commented on src/main.rs:
```diff
@@ -82,6 +82,18 @@ async fn main() -> Result<()> {
         );
     }
 
+    if args.include_robots {
+        add_provider(
+            &args,
+            &network_settings,
+            &mut providers,
+            &mut provider_names,
+            "Robots.txt".to_string(),
+            RobotsProvider::new,
+        );
+    }
```
No modification plan

hahwul commented on src/providers/robots.rs:
```diff
@@ -0,0 +1,117 @@
+                if line.starts_with("Disallow:") {
+                    if let Some(path) = line.strip_prefix("Disallow:").map(|s| s.trim()) {
+                        if !path.is_empty() && path != "/" {
+                            let url = format!("https://{}{}", domain, path);
+                            println!("RobotsProvider returned URL: {}", url);
+                            urls.push(url);
```
No modification plan

Adesoji1 commented:
> @Adesoji1 Thank you so much for sending the PR! First off, I moved the declaration in the code to show it under Provider Options and added help_heading.
> 
> ![](https://private-user-images.githubusercontent.com/13212227/434401144-c166d641-5bc9-492a-b465-fcb4f94aae05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjE2NTcsIm5iZiI6MTc0NDgyMTM1NywicGF0aCI6Ii8xMzIxMjIyNy80MzQ0MDExNDQtYzE2NmQ2NDEtNWJjOS00OTJhLWI0NjUtZmNiNGY5NGFhZTA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE2MzU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxNDhmYjJjZmMzNGU4YTFkZmI5ODk3MjYzNDY5MGIwZmFmMGIxMWI2MjU1ZTBmMDUyMGE0Y2QxMzE0YzIxNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GLn6UpD8fYOqncxVe9JWuNeCGtcE9WziC8s8XtgYdAo)
> 
> The functionality works well during testing, but in my environment, I'm seeing a clipping issue with the progress bar right from the start. I'll look into this tomorrow!
> 
> ![](https://private-user-images.githubusercontent.com/13212227/434401241-41dc9d36-29bb-4b8e-9032-c544bcbd3a99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjE2NTcsIm5iZiI6MTc0NDgyMTM1NywicGF0aCI6Ii8xMzIxMjIyNy80MzQ0MDEyNDEtNDFkYzlkMzYtMjliYi00YjhlLTkwMzItYzU0NGJjYmQzYTk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE2MzU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2ZjViZWM5NzQ4OGRlZTNlNmIwMjc0NjgyYjY3NTk1ZjRjNTJmZjU4MGJiNGVkZDhiZWMzYzFkZjI4YTcxZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qtzDhaGUw9OX192v7FscH4NtAer3iPfCEB_h4yFASeg)

That's great, I am happy about this, what else would you need me to do?

hahwul commented:
@Adesoji1 
Awesome, glad you're happy! I've fixed the line progress_bar issue and improved the check to account for `http` when `https` is missing. Both the changelog and README have been updated. 

If you spot any issues or have suggestions for improvement, feel free to let me know. Otherwise, I'll go ahead and merge it and aim for a release before the weekend. If everything looks good, just drop an emoji or a quick "ok"! 🙌

hahwul reviewed (approved):
Cool!

Adesoji1 commented:
> @Adesoji1 Awesome, glad you're happy! I've fixed the line progress_bar issue and improved the check to account for `http` when `https` is missing. Both the changelog and README have been updated.
> 
> If you spot any issues or have suggestions for improvement, feel free to let me know. Otherwise, I'll go ahead and merge it and aim for a release before the weekend. If everything looks good, just drop an emoji or a quick "ok"! 🙌

Yes , go ahead and merge, I will start ideating again so we can add more features in the future , I am glad you appreciated my PR 😊

Adesoji1 reviewed (commented):
😃, great work, perfect
